lib/async_hooks.js

```diff
@@ -0,0 +1,478 @@
+ * node::Environment::AsyncHooks::fields_.
+ *
+ * async_uid_fields is a Float64Array() that contains the async/trigger ids for
+ * several operations. These fields are as follows:
+ *   kCurrentAsyncId: The async id of the current execution stack.
+ *   kCurrentTriggerId: The trigger id of the current execution stack.
+ *   kAsyncUidCntr: Counter that tracks the unique ids given to new resources.
+ *   kInitTriggerId: Written to just before creating a new resource, so the
+ *    constructor knows what other resource is responsible for its init().
+ *    Used this way so the trigger id doesn't need to be passed to every
+ *    resource's constructor.
+ */
+const { async_hook_fields, async_uid_fields } = async_wrap;
+// Used to change the state of the async id stack.
+const { pushAsyncIds, popAsyncIds } = async_wrap;
+// Array of all AsyncHooks that will be iterated whenever an async event fires.
+// Using var instead of (preferably const) in order to assign
+// tmp_active_hooks_array if a hook is enabled/disabled during hook execution.
+var active_hooks_array = [];
+// Track if processing hook callbacks. Used to make sure active_hooks_array
+// isn't altered in mid execution if another hook is added or removed.
+var processing_hooks = false;
+// Use to temporarily store and updated active_hooks_array if the user enables
+// or disables a hook while hooks are being processed.
+var tmp_active_hooks_array = null;
+// Keep track of the field counds held in tmp_active_hooks_array.
+var tmp_async_hook_fields = null;
+
+// Each constant tracks how many callbacks there are for any given step of
+// async execution. These are tracked so if the user didn't include callbacks
+// for a given step, that step can bail out early.
+const { kInit, kBefore, kAfter, kDestroy, kCurrentAsyncId, kCurrentTriggerId,
+    kAsyncUidCntr, kInitTriggerId } = async_wrap.constants;
+
+// Used in AsyncHook and AsyncEvent.
+const async_id_symbol = Symbol('_asyncId');
+const trigger_id_symbol = Symbol('_triggerId');
+const init_symbol = Symbol('init');
+const before_symbol = Symbol('before');
+const after_symbol = Symbol('after');
+const destroy_symbol = Symbol('destroy');
+
+// Setup the callbacks that node::AsyncWrap will call when there are hooks to
+// process. They use the same functions as the JS embedder API.
+async_wrap.setupHooks({ init,
+                        before: emitBeforeN,
+                        after: emitAfterN,
+                        destroy: emitDestroyN });
+
+// Used to fatally abort the process if a callback throws.
+function fatalError(e) {
+  if (e instanceof Error) {
+    process._rawDebug(e.stack);
+  } else {
+    const o = { message: e };
+    Error.captureStackTrace(o, fatalError);
+    process._rawDebug(o.stack);
+  }
+  if (process.execArgv.indexOf('--abort-on-uncaught-exception') >= 0)
+    process.abort();
+  process.exit(1);
+}
+
+
+// Public API //
+
+class AsyncHook {
+  constructor(fns) {
```

Comment:
You could probably just use `constructor({ init, before, after, destroy })` if you want